tutorials/haskell/ghc-optimization-and-fusion/tutorial.md

```diff
@@ -0,0 +1,1932 @@
+   instead using (often) just one rewrite rule.
+
+In this section we will consider `build`/`foldr` fusion system that is used
+in the `base` package and powers all the functions on lists we take for
+granted.
+
+`foldr` is a familiar function, but what is `build`? It looks like this:
+
+```haskell
+build :: (forall b. (a -> b -> b) -> b -> b) -> [a]
+build g = g (:) []
+```
+
+What does it do? Why do we need it? The purpose of `build` is to keep list
+in “delayed” form by abstracting over the `(:)` “cons” operation and empty
+list `[]` “null”. The argument of the function is another function that
+takes cons-like operation `(a -> b -> b)` and null-like “starting point” `b`
+and produces something of the same type `b`. Clearly, this is a
+generalization of function that produces list-like things.
+
+`build` just gives that general function specific `(:)` function for consing
+and `[]` as starting point and we get our list back. The following example
+is taken from Duncan Coutts' thesis called “Stream Fusion: Practical
+shortcut fusion for coinductive sequence types” (yeah, the title is hairy,
+but the text itself is easy to understand and quite interesting, I recommend
+reading the whole thing!):
+
+```haskell
+build l == [1,2,3]
+  where
+    l cons nil = 1 `cons` (2 `cons` (3 `cons` nil))
+```
+
+Now the fusion system with `build` and `foldr` has only one rule:
+
+```haskell
+foldr f z (build g) = g f z
+```
+
+How does it help to eliminate intermediate lists? Let's see, `build g`
+builds some list, while `foldr f z` goes through the list “replacing” `(:)`
+applications with `f` and empty list with `z`, in fact this is a popular
+explanation of what `foldr` does:
+
+```haskell
+foldr f z [1,2,3] = 1 `f` (2 `f` (3 `f` z))
+```
+
+With that in mind, `g` is perfectly prepared to receive `f` and `z` directly
+to deliver exactly the same result!
+
+Let's rewrite our example using `build`/`foldr` fusion system:
+
+```haskell
+map2 :: (a -> b) -> [a] -> [b]
+map2 _ []     = []
+map2 f (x:xs) = f x : map2 f xs
+{-# NOINLINE map2 #-}
+
+{-# RULES
+"map2"     [~1] forall f xs. map2 f xs               = build (\c n -> foldr2 (mapFB c f) n xs)
+"map2List" [1]  forall f.    foldr2 (mapFB (:) f) [] = map2 f
+"mapFB"    forall c f g.     mapFB (mapFB c f) g     = mapFB c (f . g)
+  #-}
+
+mapFB :: (b -> l -> l) -> (a -> b) -> a -> l -> l
+mapFB c f = \x ys -> c (f x) ys
+{-# INLINE [0] mapFB #-}
+
+foldr2 :: (a -> b -> b) -> b -> [a] -> b
+foldr2 _ b []     = b
+foldr2 f b (a:as) = foldr2 f (f a b) as
+
+{-# RULES
+"build/foldr2" forall f z (g :: forall b. (a -> b -> b) -> b -> b). foldr2 f z (build g) = g f z
+  #-}
+
+fused2 :: [Int] -> Int
+fused2 = foldr2 (+) 0 . map2 sqr
+```
+
+A few comments here:
+
+* We need `NOINLINE` on `map2` to silence warning that the `"map2"` may
+  never fire because `map2` may be inlined first. Of course `map2` won't be
+  ever inlined because it's self-recursive, but GHC can't figure that out
+  (yet).
```

Comment:
- to silence **the** warning that ~~the~~ `"map2"` may never fire because `map2` may be inlined first. Of course `map2` won't **ever** be ~~ever~~ inlined...